app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt

```diff
@@ -222,9 +225,55 @@ class TodoWidget : BaseWidgetProvider() {
         setViewVisibility(R.id.widget_overlay, View.GONE)
     }
 
-    private suspend fun WebSocketRepository.getTodosGroupedByStatus(entityId: String) =
-        getTodos(entityId)?.response?.get(entityId)?.items.orEmpty()
-            .groupBy { it.isComplete }
+    private suspend fun readData(entity: TodoWidgetEntity): TodoWidgetData {
+        val lastEntityName = entity.latestUpdateData?.entityName
+        return runCatching {
+            val integrationRepository = serverManager.integrationRepository(entity.serverId)
+            val webSocketRepository = serverManager.webSocketRepository(entity.serverId)
+            val name = integrationRepository.getEntity(entity.entityId)?.friendlyName ?: lastEntityName
+            val todos = webSocketRepository.getTodos(entity.entityId)?.response?.get(entity.entityId)?.items.orEmpty()
+
+            updateLatestData(entity.id, name, todos)
+            TodoWidgetData(
+                name = name,
+                itemsByStatus = todos.groupBy { it.isComplete },
```

Comment:
You can create a small function that you can reuse or even handle this in a secondary ctor of `TodoWidgetData`